File: backend/src/schemas/withdraw.py
# @ai-generated
"""
用户佣金提现 Pydantic DTO 定义
统一入参/出参校验模型，字段与 ORM Model 对齐
- 金额统一 Decimal，禁止 float（避免精度污染）
- 提现状态枚举 WithdrawStatus 来源 constants.py（唯一权威源）
"""
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field, field_validator

from src.config.constants import WITHDRAW_MIN_AMOUNT, WithdrawStatus


# ══════════════════════════════════════════════════════
# 用户端请求 DTO
# ══════════════════════════════════════════════════════


class WithdrawApplyRequest(BaseModel):
    """用户发起提现请求体（user_id 由 JWT/请求头提取，不在 body 内）"""

    apply_amount: Decimal = Field(
        ..., gt=0, description="申请提现金额(元)，需 >= 最低门槛 10 元"
    )

    @field_validator("apply_amount")
    @classmethod
    def validate_min_amount(cls, v: Decimal) -> Decimal:
        if v < WITHDRAW_MIN_AMOUNT:
            raise ValueError(f"最低提现金额 {WITHDRAW_MIN_AMOUNT} 元")
        return v


# ══════════════════════════════════════════════════════
# 后台审核请求 DTO
# ══════════════════════════════════════════════════════


class WithdrawApproveRequest(BaseModel):
    """审核通过请求体（review_user_id 由 JWT 提取）"""

    review_remark: str = Field(default="", max_length=512, description="审核备注")


class WithdrawRejectRequest(BaseModel):
    """驳回请求体（驳回原因必填，用于 reject_reason 落库 + 用户可见）"""

    reject_reason: str = Field(..., min_length=1, max_length=512, description="驳回原因")


class WithdrawCompleteRequest(BaseModel):
    """标记打款完成请求体（绑定微信转账批次ID，触发冻结释放）"""

    transfer_batch_id: str = Field(
        ..., min_length=1, max_length=64, description="微信转账批次ID"
    )


class WithdrawFailRequest(BaseModel):
    """打款失败退回请求体（APPROVED/PROCESSING → REJECTED，退回可用余额）

    场景：微信转账失败 / 银行卡异常（V2.0：资金退回可用余额，后台留痕）。
    """

    reason: str = Field(..., min_length=1, max_length=512, description="打款失败原因")


# ══════════════════════════════════════════════════════
# 后台列表查询参数 DTO
# ══════════════════════════════════════════════════════


class WithdrawAdminListQuery(BaseModel):
    """后台提现申请列表查询参数"""

    user_id: Optional[int] = Field(default=None, gt=0, description="平台用户ID筛选")
    status: Optional[str] = Field(default=None, description="提现状态筛选：PENDING/APPROVED/PROCESSING/SUCCESS/REJECTED")

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        valid = {s.value for s in WithdrawStatus}
        if v not in valid:
            raise ValueError(f"status 必须为合法值: {sorted(valid)}")
        return v

File: admin/src/api/withdraw.ts
// @ai-generated
/**
 * 提现审核后台 API
 * 对接后端路由：/api/v1/admin/withdraw（权限码：withdraw:review）
 * 字段与 src/services/withdraw_service.py 对齐
 *
 * 提现状态机：
 *   PENDING 待审核 → APPROVED 审核通过 → PROCESSING 打款中 → SUCCESS 打款成功
 *   PENDING → REJECTED 驳回 / 审核失败
 *   APPROVED/PROCESSING → REJECTED 打款失败退回
 */
import request from '@/utils/request'

/** 提现状态码 */
export const WithdrawStatus = {
  PENDING: 'PENDING',
  APPROVED: 'APPROVED',
  PROCESSING: 'PROCESSING',
  SUCCESS: 'SUCCESS',
  REJECTED: 'REJECTED',
} as const

export type WithdrawStatusCode = (typeof WithdrawStatus)[keyof typeof WithdrawStatus]

/** 提现状态元数据 */
export const WithdrawStatusMeta: Record<
  string,
  { label: string; type: 'warning' | 'success' | 'info' | 'danger' }
> = {
  PENDING: { label: '待审核', type: 'warning' },
  APPROVED: { label: '审核通过', type: 'info' },
  PROCESSING: { label: '打款中', type: 'info' },
  SUCCESS: { label: '已到账', type: 'success' },
  REJECTED: { label: '已驳回', type: 'danger' },
}

/** 提现申请列表项 */
export interface WithdrawApplyItem {
  id: number
  user_id: number
  apply_amount: string
  fee: string
  actual_amount: string
  status: string
  bank_name: string
  bank_card_no: string
  apply_time: string | null
  review_time: string | null
  review_user_id: number | null
  review_remark: string
  reject_reason: string
  transfer_batch_id: string
  fail_time: string | null
  fail_reason: string
  create_time: string | null
  update_time: string | null
}

/** 提现列表响应 */
export interface WithdrawListResponse {
  list: WithdrawApplyItem[]
  total: number
  page: number
  page_size: number
}

/** 提现列表查询参数 */
export interface WithdrawListParams {
  page: number
  page_size: number
  user_id?: number
  status?: string
}

/** 审核通过请求 */
export interface WithdrawApproveRequest {
  review_remark?: string
}

/** 驳回请求 */
export interface WithdrawRejectRequest {
  reject_reason: string
}

/** 打款完成请求 */
export interface WithdrawCompleteRequest {
  transfer_batch_id: string
}

/** 打款失败退回请求 */
export interface WithdrawFailRequest {
  reason: string
}

export const withdrawApi = {
  /** 提现申请列表（后台审核工作台） */
  list(params: WithdrawListParams) {
    return request.get<WithdrawListResponse>('/v1/admin/withdraw/applies', { params })
  },

  /** 审核通过（PENDING → APPROVED） */
  approve(applyId: number, data: WithdrawApproveRequest) {
    return request.put(`/v1/admin/withdraw/applies/${applyId}/approve`, data)
  },

  /** 驳回（PENDING → REJECTED） */
  reject(applyId: number, data: WithdrawRejectRequest) {
    return request.put(`/v1/admin/withdraw/applies/${applyId}/reject`, data)
  },

  /** 标记打款完成（APPROVED/PROCESSING → SUCCESS） */
  complete(applyId: number, data: WithdrawCompleteRequest) {
    return request.put(`/v1/admin/withdraw/applies/${applyId}/complete`, data)
  },

  /** 打款失败退回（APPROVED/PROCESSING → REJECTED） */
  fail(applyId: number, data: WithdrawFailRequest) {
    return request.put(`/v1/admin/withdraw/applies/${applyId}/fail`, data)
  },
}
